cpu/stm32/periph/backup_ram.c

```diff
@@ -0,0 +1,121 @@
+ *         backup RAM content during standby mode
+ * @{
+ */
+#if defined(PWR_CSR1_BRE)
+/* So far only F7 has CSR1 */
+#define PWR_BRE_REG         PWR->CSR1
+#define BKPRAM_CONFIG       (PWR_CSR1_BRE | PWR_CSR1_EIWUP)
+#define BKPRAM_READY        (PWR_CSR1_BRR)
+#else
+/* F2, F4 */
+#define PWR_BRE_REG         PWR->CSR
+#define BKPRAM_CONFIG       (PWR_CSR_BRE)
+#define BKPRAM_READY        (PWR_CSR_BRR)
+#endif
+/** @} */
+
+/**
+ * @name   Constants depending on CPU family
+ * @{
+ */
+#if defined(CPU_FAM_STM32F7) || defined(CPU_FAM_STM32F4)
+#define VBAT_ADC_SCALE      4
+#define VBAT_ADC_MIN        1650            /* 1.65V <= VBAT <= 3.6V */
+#elif defined(CPU_FAM_STM32F2)
+#define VBAT_ADC_SCALE      2
+#define VBAT_ADC_MIN        1800            /* 1.8V <= VBAT <= 3.6V */
+#endif
+
+#define VBAT_ADC_RES        ADC_RES_12BIT   /* 12 bit fit all */
+/** @} */
+
+static inline bool _backup_battery_connected(void) {
+#if defined(VBAT_ADC) && IS_USED(MODULE_PERIPH_ADC)
+    int32_t sample = adc_sample(VBAT_ADC, VBAT_ADC_RES);
+    sample *= VBAT_ADC_SCALE;
+    return sample >= VBAT_ADC_MIN;
+#endif
+    /* if the board does not expose a VBAT pin or ADC is not implemented,
+       you can define BOARD_HAS_BACKUP_BATTERY and promise that
+       you have correctly soldered a working battery,
+       else backup ram content is undefined */
+    return IS_ACTIVE(BOARD_HAS_BACKUP_BATTERY);
+}
+
+void backup_ram_init(void)
+{
+    /* see reference manual 4.1.5 "Battery backup domain" */
+    periph_clk_en(APB1, RCC_APB1ENR_PWREN);
+    stmclk_dbp_unlock();
+    periph_clk_en(AHB1, RCC_AHB1ENR_BKPSRAMEN);
+#if defined(VBAT_ADC) && IS_USED(MODULE_PERIPH_ADC)
+    adc_init(VBAT_ADC);
+#endif
+    if (_backup_battery_connected()) {
+        /* in case the board has a backup battery the regulator must be on
+           to mitigate (unexpected) outage of VDD, so RTC register and
+           backup domain register contents are not lost */
+        backup_ram_regulator_on();
+    }
+}
+
+bool backup_ram_regulator_is_on(void)
+{
+    return (PWR_BRE_REG & BKPRAM_READY) == BKPRAM_READY;
+}
+
+void backup_ram_regulator_on(void)
+{
+    assert(!stmclk_dbp_is_locked());
+    /* switch on regulator to retain backup SRAM content while in standby or VBAT mode */
+    PWR_BRE_REG |= BKPRAM_CONFIG;
+    while (!(PWR_BRE_REG & BKPRAM_READY));
+}
+
+void backup_ram_regulator_off(void)
+{
+    assert(!stmclk_dbp_is_locked());
+    /* switch off regulator to save power, unless the board has a backup battery (VBAT)
+       to retain RTC register and backup domain register contents when VDD is not present */
+    if (!_backup_battery_connected()) {
+        PWR_BRE_REG &= ~BKPRAM_CONFIG;
+    }
+}
+
+bool cpu_woke_from_backup(void) {
+    if (_backup_battery_connected() && backup_ram_regulator_is_on()) {
+        /* if regulator is on and a battery is connected after reset,
+           backup ram should have been retained */
+        return true;
```

Comment:
Maybe that function name was a bad idea since originally it should just return if we did a cold boot or woke from Deep Sleep, but you are right that for *this* use case we also should know if the backup battery was connected.
But shouldn't this then return `false` if the backup battery is _not_ connected()?

Then again if no backup battery exists we can't use the backup RAM and there is little the software can do about it, so we can just ignore it. Or are there cases where backup RAM retention works without backup battery?

Reply:
The battery is used to retain backup ram when there is no power supply, E.g. someone removed the USB cable. If there is no battery and the board entered standby mode while the power supply was not interrupted, the backup ram is retained even without battery.

Reply:
Or maybe rewrite like this?:
```C
    if (backup_ram_regulator_is_on()) {
        if (_backup_battery_connected()) {
            /* if regulator is on and a battery is connected after reset,
            backup ram should have been retained */
            return true;
        }
        uint32_t coldboot = (RCC->CSR & (RCC_CSR_PORRSTF | RCC_CSR_BORRSTF));
        RCC->CSR |= RCC_CSR_RMVF; /* clear reset flags */
        return !(coldboot == (RCC_CSR_PORRSTF | RCC_CSR_BORRSTF));
    }
    return false;
```

Reply:
> So backup_ram_regulator_is_on() is only true if we did not cold boot?

If we enter the first `if` and skip the second `if`, `coldboot` can still be true or false, so no.
What I did not mention so far is a note in the manual regarding the `BRE` and `BRR` bits (the regulator).
> Note: This bit is not reset when the device wakes up from Standby mode, by a system reset, 
or by a power reset. [RM0410 p. 146] 

I try to formulate my thoughts more verbose.
This function answers the question: `Does the backup ram contain valid data (true) or can it be overridden.`
We are at the point where the function is called, the reset handler, so right after reset.
A necessary condition for the backup ram to be valid after reset is that the regulator must have been on.
So if that is not true, it returns `false`.
If the regulator was on and we have a battery, then backup ram survived even cold boots, so that is not tested further and it returns `true`.
Else backup ram survived only soft resets where the power supply was not interrupted.

By the time of writing I notice two things.
1. `backup_ram_init()` possibly (re)enables the regulator, thus checking whether the regulator was on is meaningless. :hushed:
2. No other function after that one will be able to check the reset flags because they are cleared immediately. A good place to clear reset flags would be at the and of `reset_handler_default()`, I guess.
But this is architecture dependent. Could this be done in an `stm32::post_startup()`, or a new `reset_done()` at the very end?

```C
void backup_ram_init(void)
{
    /* see reference manual 4.1.5 "Battery backup domain" */
    periph_clk_en(APB1, RCC_APB1ENR_PWREN);
    stmclk_dbp_unlock();
    periph_clk_en(AHB1, RCC_AHB1ENR_BKPSRAMEN);
#if defined(VBAT_ADC) && IS_USED(MODULE_PERIPH_ADC)
    adc_init(VBAT_ADC);
#endif
}

bool cpu_woke_from_backup(void) {
    bool regon = backup_ram_regulator_is_on();
    if (_backup_battery_connected()) {
        if (regon) {
            /* if regulator is on and a battery is connected after reset,
            backup ram should have been retained */
            return true;
        }
        /* in case the board has a backup battery the regulator must be on
        to mitigate (unexpected) outage of VDD, so RTC register and
        backup domain register contents are not lost */
        backup_ram_regulator_on();
        return false;
    }
    uint32_t coldboot = (RCC->CSR & (RCC_CSR_PORRSTF | RCC_CSR_BORRSTF));
    RCC->CSR |= RCC_CSR_RMVF; /* clear reset flags somewhere else */
    return regon && !(coldboot == (RCC_CSR_PORRSTF | RCC_CSR_BORRSTF));
}
```

Reply:
I am sorry for the confusion. I think now the battery + regulator thing is clear. Of course I am open for change requests as long as it works out.